eth/downloader/whitelist/service.go

```diff
@@ -70,6 +72,96 @@ func (w *Service) IsValidChain(remoteHeader *types.Header, fetchHeadersByNumber
 	return false, ErrCheckpointMismatch
 }
 
+// IsValidChain checks the validity of chain by comparing it
+// against the local checkpoint entries
+func (w *Service) IsValidChain(currentHeader *types.Header, chain []*types.Header) bool {
+	var (
+		oldestCheckpointNumber uint64 = w.checkpointOrder[0]
+		first                  uint64 = chain[0].Number.Uint64()
+		last                   uint64 = chain[len(chain)-1].Number.Uint64()
+		current                uint64 = currentHeader.Number.Uint64()
+	)
+
+	// Check if we have checkpoints to validate incoming chain in memory
+	if len(w.checkpointWhitelist) == 0 {
+		// We don't have any entries, no additional validation will be possible
+		return true
+	}
+
+	// Check if we have whitelist entries in required range
+	if last < oldestCheckpointNumber {
+		// We have future whitelisted entries, so no additional validation will be possible
+		// This case will occur when bor is in middle of sync, but heimdall is ahead/fully synced.
+		return true
+	}
+
+	// Start iterating over the chain and apply validation for past and future
+	// chain, referring to chain behind and after current header respectively.
+	for i := 0; i < len(chain); i++ {
+	}
+
+	var (
+		pastChain   []*types.Header
+		futureChain []*types.Header
+	)
+	if current >= first {
+		if len(chain) == 1 {
+			pastChain = chain
+		} else {
+			pastChain = chain[:current-first]
+		}
+	}
+	if current < last {
+		if len(chain) == 1 {
+			futureChain = chain
+		} else {
+			futureChain = chain[current-first+1:]
+		}
+	}
+
+	if len(futureChain) > int(w.checkpointInterval) {
+		return false
+	}
+
+	// Iterate over the chain and validate against the last checkpoint
+	// It will handle all cases where the incoming chain has atleast one checkpoint
+	for i := len(pastChain) - 1; i >= 0; i-- {
+		if _, ok := w.checkpointWhitelist[pastChain[i].Number.Uint64()]; ok {
+			if pastChain[i].Hash() != w.checkpointWhitelist[pastChain[i].Number.Uint64()] {
+				return false
+			} else {
+				// No need to check for other checkpoints (if any) as the latest checkpoint is valid
+				break
+			}
+		}
+	}
+
+	return true
+}
+
+func (w *Service) GetCheckpoints(current, sidechainHeader *types.Header, sidechainCheckpoints []*types.Header) (map[uint64]*types.Header, error) {
+	// todo: carefully check every possible case :)
+	return map[uint64]*types.Header{}, nil
+}
+
+func (w *Service) GetCheckpointsBetween(from, to uint64) []uint64 {
+	if from > to {
+		return nil
+	}
+
+	capacity := (to-from)/w.checkpointInterval + 1
+	res := make([]uint64, 0, capacity)
+
+	for n := from; n <= to; n++ {
+		//todo: check if it's correct
+		if n%w.checkpointInterval == 0 {
```

Comment:
I believe we already have function like IsCheckpoint somewhere in the code. If not, we need to introduce one and remove code duplication.

Reply:
I believe we can remove the functions `GetCheckpoints` and `GetCheckpointsBetween` from the whitelist service as they're no longer used. IsCheckpoint check I believe is just checking against the whitelist map, right? Should we also add it into function?